.github/workflows/build-package.yml

```diff
@@ -87,94 +87,84 @@ env:
   VERSION: ${{ inputs.version }}
   FEATURES: ${{ inputs.features }}
   NO_DEFAULT_FEATURES: ${{ inputs.NO_DEFAULT_FEATURES }}
-  BUILD_DIR: ${{ github.workspace }}/built.d
 
 jobs:
+  check-if-already-built:
+    if: inputs.always_build == ''
+    name: Check if the package ${{ github.event.inputs.crate }}-${{ github.event.inputs.version }}-${{ github.event.inputs.target_arch }} is already built
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check if the package is built
+        env:
+          USER_AGENT: cargo-quickinstall build pipeline (alsuren@gmail.com)
+        run: |
+          if curl --user-agent "$USER_AGENT" --fail -I --output /dev/null "${GITHUB_REPOSITORY}/releases/download/${CRATE}-${VERSION}/${CRATE}-${VERSION}-${TARGET_ARCH}.tar.gz"; then
+            echo "${CRATE}/${VERSION}/${CRATE}-${VERSION}-${TARGET_ARCH}.tar.gz already uploaded. Skipping."
+            echo "NEEDS_BUILD=false" >> "$GITHUB_OUTPUT"
+          else
+            echo "NEEDS_BUILD=true" >> "$GITHUB_OUTPUT"
+          fi
+          
   build-popular-package:
-    name: build-popular-package-${{ inputs.crate }}-${{ inputs.version }}-${{ inputs.target_arch }}-${{ inputs.features }}
+    needs: check-if-already-built
+    if: inputs.always_build != '' || needs.check-if-already-built.outputs.NEEDS_BUILD == 'true'
+    name: Build package ${{ inputs.crate }}-${{ inputs.version }}-${{ inputs.target_arch }}-${{ inputs.features }}
     runs-on: ${{ inputs.build_os }}
     permissions: {}
```

Comment:
(splitting my reply into line comments so that we can have multiple threads, because there's a lot going on)

> Otherwise any action written in JS won't have access to the filesystem in the VM and won't be able to upload anything, and AFAIK action written in JS just import fs as normal, so I doubt there's any remote FS access there.
> 
> Since the action needs to upload data, it must have access to the secrets.GITHUB_TOKEN, which is a bit dangerous:
>
> At the very least, having access to it allow attackers to launch a DDOS attack using the GitHub token, prevent any job from running by exhausting the points of the restful/graphql API.

My understanding was that it should not have enough scopes to do anything, after I set this option. You are right that the actions/upload-artifact action clearly manages to use **some** kind of github api to do its uploading. It is worth reading into how that works, and seeing if we can replicate it using something like:

```yaml
- name: Setup tmate session
  uses: mxschmitt/action-tmate@v3
  with:
    limit-access-to-actor: true
```

to drop us into a tmux session on the runners, so we can fuck around in bash.

Reply:
if it is possible to do anything non-job-local (e.g. writing to the @actions/cache cache, or burning our repo-wide github api rate limits) from a tmate session in a job with `permissions: {}` then we might need to pause the builders for a bit until we can do a proper audit of how we're doing things

Reply:
https://github.com/actions/toolkit/blob/361a115e538ac6d8eb06cc47f3fcecce557d04c8/packages/artifact/src/internal/shared/config.ts#L10

It seems that the github token is passed to the JS action as environment variable

Reply:
Okay, so the next question is how much can that token actually do? 

The artifact storage seems to be sandboxed per run or something, and actions/cache is per branch, with the ability to read from your base branch's cache but not write to it (unless you set something in your permissions, here). In our case I think all builds on the same arch happen from the same branch, so I guess we should treat anything we read from the cache as untrusted? I had previously assumed that having permissions: {} meant that you couldn't write to the cache at all, so we might be safe there, as long as we don't merge your docker cache stuff (see other thread)

Reply:
They can definitely keep issuing new restful/graphql API until it reaches rate limit, in that case, all our jobs would be unable to run due to exhaustion of rate limit.

`secrets.GITHUB_TOKEN` has a primary rate limit of 1000 per hour, the secondary rate limit of 100 concurrent API requests, once it is exhaustes all other jobs could stop working.

https://docs.github.com/en/graphql/overview/rate-limits-and-node-limits-for-the-graphql-api#secondary-rate-limits

Reply:
Right, but this is only for the duration of the build, so it's only going to affect the packages that happen to be building for other architectures at the same time, until this job times out (1h?).

If you want to DoS the quickinstall workers, you could just publish 1000 empty packages on crates.io and then request them all from the stats server.

I don't think that the consequences of this attack justify the complexity of the mitigation, especially if the mitigation does not affect windows and macos.

Don't get me wrong: I am learning a lot from this PR, and it's interesting to explore the possibilities for sandboxed builds. I am just not convinced that the threat model has been thought through enough to validate that this mitigation is necessary+sufficient.

I'm happy to jump on a call this weekend and chat things through (send me a time in Europe/London timezone and I'll set up a call). I've also just handed in my notice to quit at work, so I might have some time off between jobs in a few weeks, for daytime pairing.

Reply:
I'm a bit busy at weekends and in fact, have more time during Mondays to Thursdays to discuss about this.
I'm in Sydney, so there's about 9 hours of difference.